### PR TITLE
Upgrade Playwright to 1.39.0

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Upgrade Playwright to 1.39.0

## What's new?
* [locatorAssertions.toBeAttached()](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-be-attached).
* [locator.and()](https://playwright.dev/docs/api/class-locator#locator-and) and [locator.or()](https://playwright.dev/docs/api/class-locator#locator-or).
* UI Mode now shows steps, fixtures, and attachments.
* Requires Node.js >= 16.

## Testing Instructions
CI checks are green.

### CODE REVIEW

**Comment by Mamaduka** [comment]
There is an odd issue with failing tests. Playwright now sees `role=presentation` as `role=img`.

Update: I submitted the bug report https://github.com/microsoft/playwright/issues/26809.

**Comment by Mamaduka** [comment]
I just saw that the Playwright issue was removed from the next release milestone.

While the reported case is a regression, we should also update our tests to unblock the updates.

## Why?

A spinner in the image block markup means the media is still uploading. The tests need to wait until the upload process is finished.

**Comment by Mamaduka** [comment]
I've got a solution for the image/svg problem but discovered a few more blockers from upgrading to the latest Playwright version.

## Blockers

* `@wp/scripts` - We need to update [versions](https://github.com/WordPress/gutenberg/blob/24add081d2214b3893ef47dc4e491619266052e7/packages/scripts/package.json#L75) in this package, but `playwright-core` no longer [exposes](https://github.com/microsoft/playwright/pull/22612) `cli` and script fails.
* I'm getting an odd type of script error when trying to build or commit changes. See the screenshot below.


## Screenshot
CleanShot 2023-10-13 at 10 25 22

**Comment by kevin940726** [comment]
@Mamaduka Hmm, I can't reproduce the TS errors in the screenshots. I checked out to the branch, did `npm install` and run `npm run build`, but everything builds correctly. Is there something I'm missing?

**Comment by Mamaduka** [comment]
Sorry, @kevin940726. The repro required rebase and bumping the version.

I committed changes using the `--no-verify` flag, so now you should be able to reproduce them.

1. `git pull` on the branch.
2. `npm install`.
3. `npm run build`

**Comment by kevin940726** [comment]
The API and its signature was changed in https://github.com/WordPress/gutenberg/pull/48286, but I don't think it's relevant in that PR. Possibly an older attempt to fix some issues, @ellatrix? I think we can revert that change here (done in [299a2fa](https://github.com/WordPress/gutenberg/pull/54051/commits/299a2fa9b5b73f66a71dd8bc9079d6732ff20102)), the API isn't used anywhere anyway.

**Comment by Mamaduka** [comment]
Well, that was fast 😄 Thank you, @kevin940726!

File: packages/scripts/scripts/test-playwright.js [hunk 1]
```diff
@@ -27,7 +28,10 @@ const {
 
 const result = spawn(
 	'node',
-	[ require.resolve( 'playwright-core/cli' ), 'install' ],
+	[
+		path.resolve( require.resolve( 'playwright-core' ), '..', 'cli.js' ),
+		'install',
+	],
```
Comment (Mamaduka) [inline packages/scripts/scripts/test-playwright.js:34]: @swissspidy, @kevin940726, I had to change the CLI path resolution to this (reason #22612).

Comment (swissspidy) [reply packages/scripts/scripts/test-playwright.js:34]: If it works it works :)

**Comment by Mamaduka** [comment]
Pushed the fix for the image block e2e tests (https://github.com/WordPress/gutenberg/pull/54051#issuecomment-1698886612), unless there're any other failures this should be good to merge.

File: packages/scripts/scripts/test-playwright.js (hunk 1 above)
Comment (kevin940726) [reply packages/scripts/scripts/test-playwright.js:34]: Nit: Could we just use `npx`?

**Review by kevin940726 — APPROVED** [review]
Awesome! Thank you!

File: packages/scripts/scripts/test-playwright.js (hunk 1 above)
Comment (Mamaduka) [reply packages/scripts/scripts/test-playwright.js:34]: I tried that when I initially found the issue, but it failed locally for me. Pascal might have more details on why we're using `node` vs. `npx`.

**Comment by Mamaduka** [comment]
@kevin940726, the [299a2fa](https://github.com/WordPress/gutenberg/pull/54051/commits/299a2fa9b5b73f66a71dd8bc9079d6732ff20102) is probably causing `should allow dropping an image on an empty paragraph block` test to fail.

The test was using `drop` with locator argument:

https://github.com/WordPress/gutenberg/blob/63408008fae5abf65286a8d969c4f9da98f9b86c/test/e2e/specs/editor/blocks/paragraph.spec.js#L86-L88

Update: The `test/e2e/specs/editor/various/draggable-blocks.spec.js` tests also have two similar cases.

**Comment by kevin940726** [comment]
Oops! Yep, I accidentally ignored `test` in my search 🤦. I'll push a fix soon!

**Comment by kevin940726** [comment]
Turns out the change was needed in https://github.com/WordPress/gutenberg/pull/4828 (sorry false alarm @ellatrix!) But I think there's a better fix that better matches what I had in mind with that API.

**Comment by Mamaduka** [comment]
> But I think there's a better fix that better matches what I had in mind with that API.

Is this the current version of the API or something planned for the future?

**Comment by kevin940726** [comment]
> Is this the current version of the API or something planned for the future?

The current API. The idea is that `dragOver` already has to specify an element, so `drop` shouldn't need it to avoid duplication. And also, sometimes the drop target isn't accessible so can't be targeted (like maybe it's a popover on top of the element).  It would be best to be able to just leverage `page.mouse.move` though but it's trickier.

**Comment by Mamaduka** [comment]
The [0864852](https://github.com/WordPress/gutenberg/pull/54051/commits/0864852f2dcd0b3ebb88ab9dfec423410e1c1811) commit fixes failing `should select with shift + click` test. It ran in legacy canvas mode, but the `assertNativeSelection` expects the iframed canvas. I'm unsure why this is passing on the trunk since this is definitely a bug.

cc @WunderBart

**Comment by youknowriad** [comment]
Is this PR responsible of the performance metric jump in the "first block" metric in http://codevitals.run
I wonder why? since the runner is updated for both previous and current tested commit, and the numbers are normalized. Makes me wonder if we're doing something wrong in the tests.

**Comment by Mamaduka** [comment]
> Is this PR responsible of the performance metric jump in the "first block" metric

Do you mean "jump" in a good way? If I'm reading the site correctly, the "First block load" metrics have been improved.

> I wonder why?

In theory, improved performance by the Playwright can improve our metrics. But knowing sure reason would be better. Unfortunately, I don't have a clue at the moment; maybe others have some ideas 😅

**Comment by youknowriad** [comment]
> Do you mean "jump" in a good way? If I'm reading the site correctly, the "First block load" metrics have been improved.

It goes in the right direction yeah but a change in the test tool shouldn't impact the metric ideally.

> In theory, improved performance by the Playwright can improve our metrics. But knowing sure reason would be better. Unfortunately, I don't have a clue at the moment; maybe others have some ideas 😅

Upgrading playwright can indeed make the tests faster but it shouldn't improve the metrics in our graphs because our graphs compare the current commit with a reference commit and normalize the numbers relatively to and old number for a reference commit. This means that upgrading playwright should in theory update both commits similarly and with the normalization applied, the number should stay stable in the graphs. What happens here, is that the upgrade has made the test faster for the latest commit but the upgrade didn't have the same impact on the reference commit which is strange unless we may be doing something wrong in our tests.

**Comment by WunderBart** [comment]
> What happens here, is that the upgrade has made the test faster for the latest commit but the upgrade didn't have the same impact on the reference commit which is strange unless we may be doing something wrong in our tests.

It might be an issue with the base commit. Looking at the upgrade commit job, we see the following numbersfor the `firstBlock` metric:

- Base commit: 8230.65 ms,
- Merge commit: 6252.65 ms.

Now, looking at the preceding commit, the numbers were:

- Base commit: 16681.9 ms,
- Merge commit: 14901.7 ms.

In both cases, the difference between base and merge commits is ~2 seconds. This might be a static bump between the commits independent of the framework version (or even the framework itself). The UI execution speed could have improved proportionally, but a request might have improved between the base and newer commits that will always have the same ~2s timing difference. Does that make sense?

**Comment by youknowriad** [comment]
@WunderBart it makes sense and it confirms for me that the metric computation is wrong. It feels like the metric includes a static value that comes from the test itself and not from the actual code loading.

**Comment by WunderBart** [comment]
> @WunderBart it makes sense and it confirms for me that the metric computation is wrong. It feels like the metric includes a static value that comes from the test itself and not from the actual code loading.

@youknowriad, I think the issue might be with [how we chose the new base commit recently](https://github.com/WordPress/gutenberg/pull/54111). The time between the merge commit (that updated the base) and the new base commit was almost 3 months, so there already was a big difference in `firstBlock` metric (2834ms vs 6764ms) that we didn't track / account for. The aforementioned static 2 seconds were probably already there, which broke the value normalization from day 1.

If we update/reset the base commit, shouldn't we always choose the one that is as close as possible to the merge one so there's no performance difference between the two? The normalization factor would be 1, so we'd likely see some bump from the drift itself, but from that point on, any performance change should be tracked correctly, if I'm not mistaken.

**Comment by youknowriad** [comment]
@WunderBart I'm not following, I'm not sure how this relates to the base commit at all. We're updating playwright and we're updating it for both the base commit and this commit, so the numbers should change relatively in the same way as if we didn't update playwright.

**Comment by WunderBart** [comment]
> @WunderBart I'm not following, I'm not sure how this relates to the base commit at all. We're updating playwright and we're updating it for both the base commit and this commit, so the numbers should change relatively in the same way as if we didn't update playwright.

What if they did update relatively, but there already was a 2s static bump between the merge and base commits? For example, when we re-set the base, the X metric values were:

- merge commit: 5000ms
- base commit: 7000ms
- display value: 5000 * 7000/7000 = 5000ms

There already was a 2s bump, which we don't know the source of. For the sake of this example, let's assume it came from an improved request, making it a static bump between the commits.

Let's assume the next commit updates Playwright, and the browser execution speed is 30% faster. Assuming the 2000ms is static and not affected by the update, the new values would become:

- merge commit: 5000 * 0.7 = 3500ms
- base commit: (7000-2000) * 0.7 + 2000 = 5500ms 
- display value: 3500 * 7000/5500 = 4454ms ❌ (the value was affected)

If we had re-set the base with a commit with the same performance, and even if the static bump happened along the way, the Playwright update would not affect the metric. Consider the following measurement timeline:

1. We reset the base commit
   - merge commit: 5000ms
   - base commit: 5000ms
   - display value: 5000ms
2. A static 2s bump is introduced
   - merge commit: 3000ms
   - base commit: 5000ms
   - display value: 3000 * 5000/5000 = 3000ms
3. We update Playwright (30% speed-up):
   - merge commit: 3000 * 0.7 = 2100ms
   - base commit: 5000 * 0.7 = 3500ms
   - display value: 2100 * 5000/3500 = 3000ms ✅ (the value was not affected)

I hope it makes sense! 😅

**Comment by youknowriad** [comment]
> Assuming the 2000ms is static and not affected by the update, the new values would become:

This is the assumption that I disagree with I think. If 2000ms is a "static" update, then we're measuring the wrong thing. We're not measuring the time the server loads the first block, we're measuring the time it takes playwright to load the first block. In other words, if we're measuring the correct thing, the difference should never be static across two job runs. Maybe our metric is affected by playwright timeout or things like that and it shouldn't be, we should measure real DOM timeouts, real server load times... and playwright's own code shouldn't affect the measure.

**Comment by WunderBart** [comment]
> Maybe our metric is affected by playwright timeout or things like that and it shouldn't be, we should measure real DOM timeouts, real server load times... and playwright's own code shouldn't affect the measure.

The test itself is very simple as it just waits for the first block selector (inside the editor iframe) - not much to be changed there. I'd love to confirm whether Playwright has anything to do with it, though, but I wasn't able to repro locally so far. Regardless, I think the next time we reset the base, we should choose a commit closest to the merge one so it's at least easier to track down weird bumps like this one.

**Comment by youknowriad** [comment]
playwright have built-in "waits" when you click buttons and perform any action basically. I think for performance tests, these should be disabled as much as we can.

**Comment by WunderBart** [comment]
> playwright have built-in "waits" when you click buttons and perform any action basically. I think for performance tests, these should be disabled as much as we can.

No UI actions are performed for the loading test (which includes `firstBlock`), but we do use `waitFor()` when waiting for the first block locator. The default element state for that method is [`visible`](https://playwright.dev/docs/api/class-locator#locator-wait-for:~:text=%27visible%27%20%2D%20wait%20for%20element%20to%20have%20non%2Dempty%20bounding%20box%20and%20no%20visibility%3Ahidden.%20Note%20that%20element%20without%20any%20content%20or%20with%20display%3Anone%20has%20an%20empty%20bounding%20box%20and%20is%20not%20considered%20visible.), so we can try changing it to [`attached`](https://playwright.dev/docs/api/class-locator#locator-wait-for:~:text=%27attached%27%20%2D%20wait%20for%20element%20to%20be%20present%20in%20DOM). It depends on what we mean by the block being loaded, though.

As for, e.g., a button _waits_, Playwright ensures that the button is visible, stable (not moving), and clickable (not disabled, not overlayed) before clicking it. We were also doing these checks with Puppeteer (e.g. via `waitForSelector`) – they just weren't implied. It makes sense to keep them as they stabilize tests like _Inserter Open_ or _Hovering Inserter Items_. Not sure how it would be better to disable those checks tbh. At least not for the current tests.

**Comment by kevin940726** [comment]
IMO, any browser launcher will impact the metrics, which is unavoidable. If we really want to minimize the impact then we might want to use `performance.mark` in our code to allow granular controls over when we want the test to end. However, I think in our use cases the delay caused by the test runner rarely matters, as we only care about the difference over time rather than the absolute duration for each run (as long as the delay is stable enough).

**Pull request merged**